### PR TITLE
network: automatically retry binding the listener socket

### LINKED ISSUE #13: Feature request: attempt to rebind listener sockets periodically, if the configured socket is not being listened to
Hey all,

it is possible to ask `velbustcp` to bind to a specific interface by specifying its IP address, for example as such:

```json
{
  "connections":[{
    "auth":false,
    "host":"192.168.110.1",
    "port":27015,
    "relay":true,
    "ssl":false
  }]
}
```

However, it is possible that at the time `velbustcp` is started that this address is not assigned to any of the interfaces quite yet. This can be something as simple as it being a WiFi address, and the WiFi connection has not been established yet, it could be an address that is (statically) assigned via DHCP (and DHCP did not respond yet), or it could be a simple service start up race condition, where the velbustcp service attempts to start up before the network interfaces are completely set up. I believe an issue also can come up if the network interface that the socket is bound to is disconnected and later reconnected.

In those instances you're prone to see a log print out like this:

```
Traceback (most recent call last):
   File "/bin/.velbustcp-wrapped", line 9, in <module>
     sys.exit(main())
   File "/lib/python3.10/site-packages/velbustcp/__main__.py", line 72, in main
     main = Main()
   File "/lib/python3.10/site-packages/velbustcp/__main__.py", line 36, in __init__
     self.__bridge.start()
   File "/lib/python3.10/site-packages/velbustcp/lib/connection/bridge.py", line 45, in start
     self.__network_manager.start()
   File "/lib/python3.10/site-packages/velbustcp/lib/connection/tcp/networkmanager.py", line 21, in start
      network.start()
   File "/lib/python3.10/site-packages/velbustcp/lib/connection/tcp/network.py", line 114, in start
     self.__bind_socket.bind(self.__options.address)
OSError: [Errno 99] Cannot assign requested address
2023-09-21 17:19:53,574 - INFO - bus - Serial connection active on port /dev/serial/by-id/usb-Velleman_Projects_VMB1USB_Velbus_USB_interface-if00
```

and continue running in a state of degraded functionality – it is connected to the velbus installation, but it does not present a server to other clients wanting to connect. It would be ideal if `velbustcp` attempted to recreate the listening socket occasionally, or on systems that support them used options like `IP_FREEBIND`.

### PR DESCRIPTION
**Pull request by nagisa**
It is possible that binding to a socket will fail at the program startup, for example because DHCP has not fetched the IP address for the interface, or the interface is down (because WiFi).

This changes the `Network` class to implement retries every 5 seconds if the initial bind has failed. Based on my experiments once the first bind succeeds, the socket will remain properly bound and working even if the interface loses the IP address or disappears entirely, so that is nice, but unknown how reliable this behaviour is. In particular my intuition was screming to me that `accept` call would raise an error which we would use to set `self.__bind_socket = None`, so that we'd know we need to rebind later.

While here I had to make some changes to the structure, for example:

* I had to replace `self.__running: bool` with `self.__stop:
threading.Event`, as that was required to have an interruptible sleepin the `__get_bound_socket` function.
* I replaced the "self-connect" with socket shutdown in order to advance `accept` loop, as there isn't a guarantee that the network was listening on `127.0.0.1` in the first place!
* I also made the SSL context creation part of the very same retry logic – now if somebody configures invalid SSL certificates, or if those certificates are made available at a later date (e.g. fetched from letsencrypt?) the velbustcp server won’t get stuck in a similar situation as it was getting stuck with socket binds.

Fixes #13

### CODE REVIEW

**Comment by nagisa** [comment]
Please run the github workflows before merging – I was not able to run the tests locally due to tox getting into an infinite loop on my system for some reason.

**Comment by nagisa** [comment]
I'm also okay with removing the last commit as it is only a minor optimization that's not really necessary to achieve the fix in this PR.

**Comment by dukeofphilberg** [comment]
> r? @pvanloo

Hi, we'll look to review this soon. Things have been busy here at the office 😄

File: src/velbustcp/lib/connection/tcp/network.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import socket
 import ssl
 import logging
+import platform
 from typing import List
```
Comment (dukeofphilberg) [inline src/velbustcp/lib/connection/tcp/network.py]:
```suggestion
from typing import List, Optional
```

File: src/velbustcp/lib/connection/tcp/network.py [hunk 2]
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
...
+        while self.is_active() and not self.__bind_socket:
+            # First, try to initialize the SSL context (as we only need to do it once and once done
+            # we won’t need to return to trying this, most likely.)
+            if self.__options.ssl and not self.__context:
+                try:
+                    context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+                    context.load_cert_chain(self.__options.cert, keyfile=self.__options.pk)
+                    self.__context = context
+                except Exception as e:
+                    self.__logger.error("Could not initialize SSL for %s: %s", self.__options.address, e)
```
Comment (dukeofphilberg) [inline src/velbustcp/lib/connection/tcp/network.py]: Shorthand to automatically add stacktrace to the logging call. Check https://docs.python.org/3/library/logging.html#logging.Logger.exception

```suggestion
                    self.__logger.exception("Could not initialize SSL for %s", self.__options.address)
```

File: src/velbustcp/lib/connection/tcp/network.py [hunk 3]
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
...
+                    IP_FREEBIND = 15
+                    sock.setsockopt(socket.SOL_IP, IP_FREEBIND, 1)
+                except Exception as e:
+                    self.__logger.debug("Could not set IP_FREEBIND for socket at %s: %s", self.__options.address, e)
 
             try:
-                client_socket, address = self.__bind_socket.accept()
+                sock.bind(self.__options.address)
+            except OSError as e:
+                self.__logger.error("Could not bind to %s: %s", self.__options.address, e)
```
Comment (dukeofphilberg) [inline src/velbustcp/lib/connection/tcp/network.py]:
```suggestion
                self.__logger.exception("Could not bind to %s", self.__options.address)
```

File: src/velbustcp/lib/connection/tcp/network.py [hunk 4]
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
...
+                self.__stop.wait(RETRY_DELAY)
+                continue
+
+            try:
+                sock.listen(0)
+            except OSError as e:
+                self.__logger.error("Could not listen on %s: %s", self.__options.address, e)
```
Comment (dukeofphilberg) [inline src/velbustcp/lib/connection/tcp/network.py]:
```suggestion
                self.__logger.exception("Could not listen on %s", self.__options.address)
```

File: src/velbustcp/lib/connection/tcp/network.py [hunk 5]
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
...
+        """Binds a listener socket and accept clients from it.
+
+        If the tcp server is closed it will also close socket
+        """
+
+        while self.is_active():
+            try:
+                client_socket, address = self.__get_bound_socket().accept()
+            except OSError as e:
+                self.__logger.error("Couldn't accept socket: %s", e)
```
Comment (dukeofphilberg) [inline src/velbustcp/lib/connection/tcp/network.py]:
```suggestion
                self.__logger.exception("Couldn't accept socket")
```

File: src/velbustcp/lib/connection/tcp/network.py [hunk 6]
```diff
@@ -59,39 +65,92 @@ def send(self, data: bytearray) -> None:
                 except Exception:
                     self.__logger.exception("Could not send data to client %s", client.address())
 
-    def __accept_sockets(self) -> None:
-        """Accepts clients from socket, if the tcp server is closed it will also close socket
-        """
+    def __get_bound_socket(self) -> socket.socket:
```
Comment (dukeofphilberg) [inline src/velbustcp/lib/connection/tcp/network.py]:
```suggestion
    def __get_bound_socket(self) -> Optional[socket.socket]:
```

If we didn't start the server yet, this call will return None

**Review by dukeofphilberg — CHANGES_REQUESTED** [review]
Hi, sorry again that this took so long to get to!

It looks and functions well, thanks for the contribution. I've added some remarks. I can run tox again through GH actions after you make these changes, the initial workflow run request has disappeared.

**Comment by nagisa** [comment]
Thanks for the review. I have applied all of the suggestions.

**Comment by dukeofphilberg** [comment]
> Thanks for the review. I have applied all of the suggestions.

I've started the tox gh action check but after 30minutes they were still running. I'm assuming the same issue you had locally. I'll check this again tomorrow.

**Comment by nagisa** [comment]
> I've started the tox gh action check but after 30minutes they were still running. I'm assuming the same issue you had locally. I'll check this again tomorrow.

In my case it was tox itself hanging, but I was able to run pytest locally and was able to reproduce the issue that GitHub hung on as well. I have pushed two commits to fix the issues pointed out by pytest.

---

The problem was that in the network.py `stop()` function we had this;

```
try:
    self.__bind_socket.shutdown(socket.SHUT_RDWR)
    self.__bind_socket.close()
    self.__bind_socket = None
except AttributeError:
    pass
```

and then 

```
self.__server_thread.join()
```

but it was possible that the `__bind_socket` gets set to a bound socket in between these two blocks, which would lead to the server thread listening on a properly functioning socket without any intent to shut down, while the `stop()` was thinking it did everything to terminate the sockets and was waiting for the server thread to shut down.

**Comment by dukeofphilberg** [comment]
> > I've started the tox gh action check but after 30minutes they were still running. I'm assuming the same issue you had locally. I'll check this again tomorrow.
> 
> In my case it was tox itself hanging, but I was able to run pytest locally and was able to reproduce the issue that GitHub hung on as well. I have pushed two commits to fix the issues pointed out by pytest.
> 
> The problem was that in the network.py `stop()` function we had this;
> 
> ```
> try:
>     self.__bind_socket.shutdown(socket.SHUT_RDWR)
>     self.__bind_socket.close()
>     self.__bind_socket = None
> except AttributeError:
>     pass
> ```
> 
> and then
> 
> ```
> self.__server_thread.join()
> ```
> 
> but it was possible that the `__bind_socket` gets set to a bound socket in between these two blocks, which would lead to the server thread listening on a properly functioning socket without any intent to shut down, while the `stop()` was thinking it did everything to terminate the sockets and was waiting for the server thread to shut down.


tox has run now, some errors in flake8 and mypy

**Comment by nagisa** [comment]
Sorry about that! I have fixed both flake8’s and mypy’s concerns in the last two commits.

**Comment by nagisa** [comment]
Just in case this fell through the cracks, is there any chance you could give this another look?

**Comment by dukeofphilberg** [comment]
We'll be looking to get this merged this week.

**Comment by dukeofphilberg** [comment]
Hi @nagisa, thanks for your patience. I've finalized the PR with some additional logic to catch errors. Thanks again!

**Pull request merged**